compiler/rustc_typeck/src/errors.rs

```diff
@@ -324,3 +324,21 @@ pub struct SubstsOnOverriddenImpl {
     #[primary_span]
     pub span: Span,
 }
+
+#[derive(LintDiagnostic)]
+#[lint(typeck::unused_extern_crate)]
+pub struct UnusedExternCrate {
+    #[primary_span]
```

Comment:
You shouldn't need `#[primary_span]` in a `LintDiagnostic`, because the call to `emit_spanned_lint` takes a `Span` (which is how we calculate whether the lint has a allow/deny/etc in that location). I believe that the test failure we're seeing is because this addition is equivalent to doing..

```rust
tcx.emit_spanned_lint(lint, id, span_with_attrs, UnusedExternCrate { /* .. */ });
```

..instead of what we currently have, which is correct:

```rust
tcx.emit_spanned_lint(lint, id, span, UnusedExternCrate { /* .. */ });
```

This is something that we should fix in the `LintDiagnostic` derive, it's an easy mistake to make and it probably isn't well documented.

Reply:
Thanks! Fixed.